java/src/main/native/ai_onnxruntime_OnnxTensor.c

```diff
@@ -136,47 +161,55 @@
  */
 JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OnnxTensor_createStringTensor
         (JNIEnv * jniEnv, jclass jobj, jlong apiHandle, jlong allocatorHandle, jobjectArray stringArr, jlongArray shape) {
-    (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
+    (void) jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
     const OrtApi* api = (const OrtApi*) apiHandle;
     OrtAllocator* allocator = (OrtAllocator*) allocatorHandle;
 
     // Extract the shape information
-    jboolean mkCopy;
-    jlong* shapeArr = (*jniEnv)->GetLongArrayElements(jniEnv,shape,&mkCopy);
-    jsize shapeLen = (*jniEnv)->GetArrayLength(jniEnv,shape);
+    jlong* shapeArr = (*jniEnv)->GetLongArrayElements(jniEnv, shape, NULL);
+    jsize shapeLen = (*jniEnv)->GetArrayLength(jniEnv, shape);
 
     // Array length
     jsize length = (*jniEnv)->GetArrayLength(jniEnv, stringArr);
 
     // Create the OrtValue
-    OrtValue* ortValue;
-    checkOrtStatus(jniEnv, api, api->CreateTensorAsOrtValue(allocator,(int64_t*)shapeArr,shapeLen,ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING,&ortValue));
-    (*jniEnv)->ReleaseLongArrayElements(jniEnv,shape,shapeArr,JNI_ABORT);
-
-    // Create the buffers for the Java strings
-    const char** strings;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(char*)*length,(void**)&strings));
-    jobject* javaStrings;
-    checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator,sizeof(jobject)*length,(void**)&javaStrings));
-
-    // Copy the java strings into the buffers
-    for (jsize i = 0; i < length; i++) {
-        javaStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv,stringArr,i);
-        strings[i] = (*jniEnv)->GetStringUTFChars(jniEnv,javaStrings[i],NULL);
-    }
-
-    // Assign the strings into the Tensor
-    checkOrtStatus(jniEnv, api, api->FillStringTensor(ortValue,strings,length));
-
-    // Release the Java strings
-    for (int i = 0; i < length; i++) {
-        (*jniEnv)->ReleaseStringUTFChars(jniEnv,javaStrings[i],strings[i]);
+    OrtValue* ortValue = NULL;
+    OrtErrorCode code = checkOrtStatus(jniEnv, api, api->CreateTensorAsOrtValue(allocator, (int64_t*)shapeArr, shapeLen,
+                                        ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING, &ortValue));
+    (*jniEnv)->ReleaseLongArrayElements(jniEnv, shape, shapeArr, JNI_ABORT);
+
+    if (code == ORT_OK) {
+      // Create the buffers for the Java strings
+      const char** strings = NULL;
+      code = checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator, sizeof(char*) * length, (void**)&strings));
```

Comment:
There is no reason to use ORT Allocator here. One can just use malloc or whatever.  

The data is internally copied anyway because C++ code stores it in std::string objects.

Reply:
I've been moderately consistently using ORT's allocator as otherwise there are three different memory allocators in use (Java's, malloc and ORT's allocators). I can replace the use of ORT's allocators with malloc if you want, but I'd assumed it was ok as ORT uses it's own allocators to allocate strings in a bunch of the C API methods, rather than malloc and requiring users to free that memory.

Reply:
I did not mean to replace all the case. Sometimes, the API requires it to be allocated with the ORT allocator. But in this case, it is just a piece of temporary memory which could be on the stack if we could do it.